oauth2client/contrib/gce.py

```diff
@@ -85,32 +115,100 @@ class AppAssertionCredentials(AssertionCredentials):
     information to generate and refresh its own access tokens.
     """
 
-    @util.positional(2)
-    def __init__(self, scope='', **kwargs):
+    @util.positional(3)
+    def __init__(self,
+                 scope=None,
+                 service_account_email='default',
+                 service_account_info=None,
+                 **kwargs):
         """Constructor for AppAssertionCredentials
 
         Args:
-            scope: string or iterable of strings, scope(s) of the credentials
-                   being requested. Using this argument will have no effect on
-                   the actual scopes for tokens requested. These scopes are
-                   set at VM instance creation time and won't change.
+            scope:
+                string or iterable of strings, scope(s) of the credentials
+                being requested. Using this argument will have no effect on
+                the actual scopes for tokens requested. These scopes are
+                set at VM instance creation time and won't change.
+            service_account_email:
+                the email for these credentials. This can be used with custom
+                service accounts, or left blank to use the default service
+                account for the instance. Usually the compute engine service
+                account.
+            service_account_info:
+                Deserialized JSON object, returned by self.service_account_info
         """
+
+        self._service_account_info = service_account_info or {
+            'email': service_account_email
+        }
+        self._project_id = None
+        self._partial = service_account_info is None
+
         if scope:
-            warnings.warn(_SCOPES_WARNING)
-        # This is just provided for backwards compatibility, but is not
-        # used by this class.
-        self.scope = util.scopes_to_string(scope)
+            if self.has_scopes(scope):
+                warnings.warn(_SCOPES_WARNING)
+            else:
+                raise ValueError(_SCOPES_WARNING)
+
         self.kwargs = kwargs
 
         # Assertion type is no longer used, but still in the
         # parent class signature.
         super(AppAssertionCredentials, self).__init__(None)
-        self._service_account_email = None
 
-    @classmethod
-    def from_json(cls, json_data):
-        data = json.loads(_from_bytes(json_data))
-        return AppAssertionCredentials(data['scope'])
+    @property
+    def service_account_info(self):
+        return self._get_service_account_info()
+
+    @property
+    def scopes(self):
+        return self.service_account_info['scopes']
+
+    @scopes.setter
```

Comment:
Ok, lemme try to lay this out more clearly:

My desired behavior:
Any time a `scope` is specified (either via `create_scoped_required`, or with a `scope` kwarg, or by assigning to the `scope` property), I want `AppAssertionCredentials` to fail fast, by checking these scopes against the immutable scopes of the credentials, and warning if they are a subset, and erroring if they have are not. This way users don't have to wait to make a call to know they won't work (and we can provide a better error message).

There are lots of weird inconsistencies in how scopes are handled in this library:
 * The constructor for `Credentials` takes a `scopes` argument which is assigned to the property `self.scopes`
* Neither `GoogleCredentials` nor `AssertionCredentials` pass unused kwargs up to their parent's constructor. So there is no way for me to get at this arg. 
* This function defines (apparently for legacy reasons) a `scope` argument. it's not clear to me why this is different from the `scopes` argument of `Credentials`, or why it needs to be provided (given that the parent class uses `scopes`)

Overall this class hierarchy has me quite confused, and I'm not sure how to implement the behavior I want, and what exactly I'm expected to support...